mutual-information-feature-selection/titanic.py

```diff
@@ -0,0 +1,155 @@
+import itertools
+import os
+
+import matplotlib.pyplot as plt
+import matplotlib.colors as colors
+import numpy as np
+import pandas as pd
+
+# D-Wave Ocean tools
+import dimod
+import neal
+from dwave.embedding.chimera import find_clique_embedding
+from dwave.system import DWaveSampler, FixedEmbeddingComposite
+
+# Read the feature-engineered data into a pandas dataframe
+# Data obtained from http://biostat.mc.vanderbilt.edu/DataSets
+datapath = os.path.join(os.path.dirname(os.path.abspath(__file__)),
+                        'data',
+                        'formatted_titanic.csv')
+dataset = pd.read_csv(datapath)
+
+
+# Define MI calculations
+def prob(dataset):
+    """Joint probability distribution P(X) for the given data."""
+
+    # bin by the number of different values per feature
+    num_rows, num_columns = dataset.shape
+    bins = [len(np.unique(dataset[:, ci])) for ci in range(num_columns)]
+
+    prob, _ = np.histogramdd(dataset, bins)
+    return prob / np.sum(prob)
+
+
+def shannon_entropy(p):
+    """Shannon entropy H(X) is the sum of P(X)log(P(X)) for probabilty
+    distribution P(X).
+    """
+    p = p.flatten()
+    return -sum(pi*np.log2(pi) for pi in p if pi)
+
+
+def conditional_shannon_entropy(p, *conditional_indices):
+    """Conditional Shannon entropy H(X;Y) = H(X,Y) - H(Y)."""
+
+    axis = tuple(i for i in np.arange(len(p.shape))
+                 if i not in conditional_indices)
+
+    return shannon_entropy(p) - shannon_entropy(np.sum(p, axis=axis))
+
+
+def mutual_information(prob, j):
+    """Mutual information between variables X and variable Y.
+
+    Calculated as I(X; Y) = H(X) - H(X|Y)."""
+
+    return (shannon_entropy(np.sum(prob, axis=j))
+            - conditional_shannon_entropy(prob, j))
+
+
+def conditional_mutual_information(p, j, *conditional_indices):
+    """Mutual information between variables X and variable Y conditional on variable Z.
+
+    Calculated as I(X;Y|Z) = H(X|Z) - H(X|Y,Z)"""
+
+    return (conditional_shannon_entropy(np.sum(p, axis=j), *conditional_indices)
```

Comment:
in `conditional_shannon_entropy(np.sum(p, axis=j), *conditional_indices)` wouldn't the conditional_indices need to be offset by one if they are larger than `j`? ie. when you marginalize over `j` in `np.sum(p, axis=j)`, the number of axes in `p` shrinks by 1

Reply:
I guess I'm wrong since this behaviour I'm describing should raise an error in lines 111 and 112 (there is no index 2 after the probability is marginalized over index 1)